ui/components/src/ui-components/inputs/ParsedInput.tsx

```diff
@@ -0,0 +1,112 @@
+  /** Function used to parse user input into a value */
+  parseString: (stringValue: string) => ParseResults;
+  /** Function to call when value is changed */
+  onChange?: (newValue: number) => void;
+  /** if readonly then only the formatValue function is used. */
+  readonly?: boolean;
+}
+
+/** Generic Input component that requires formatting and parsing functions to be passed in as props.
+ * @beta
+ */
+export const ParsedInput = React.forwardRef<HTMLInputElement, ParsedInputProps>(
+  function ParsedInput({ initialValue, formatValue, parseString, readonly, className, style, onChange }, ref) {
+    const currentValueRef = React.useRef(initialValue);
+    const isMountedRef = React.useRef(false);
+    const lastFormattedValueRef = React.useRef(formatValue(initialValue));
+    const [formattedValue, setFormattedValue] = React.useState(() => lastFormattedValueRef.current);
+    const [hasBadInput, setHasBadInput] = React.useState(false);
+
+    React.useEffect(() => {
+      isMountedRef.current = true;
+      return () => {
+        isMountedRef.current = false;
+      }
+    }, [])
+
+    // See if new initialValue props have changed since component mounted
+    React.useEffect(() => {
+      // istanbul ignore else
+      if (initialValue !== currentValueRef.current) {
+        currentValueRef.current = initialValue;
+        lastFormattedValueRef.current = formatValue(initialValue);
+        setFormattedValue(lastFormattedValueRef.current);
+        setHasBadInput(false);
+      }
+    }, [formatValue, initialValue])
+
+    const handleInputChange = React.useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
+      setFormattedValue(event.currentTarget.value);
+    }, [])
+
+    const updateQuantityValueFromString = React.useCallback((strVal: string) => {
+      if (lastFormattedValueRef.current === strVal)
+        return;
+
+      const parseResults = (parseString(strVal));
+      // istanbul ignore else
+      if (!parseResults.parseError) {
+        // istanbul ignore else
+        if (undefined !== parseResults.value) {
+          // istanbul ignore else
+          if (currentValueRef.current !== parseResults.value) {
+            currentValueRef.current = parseResults.value as number;
+            lastFormattedValueRef.current = formatValue(currentValueRef.current);
+            onChange && onChange(currentValueRef.current);
+            // istanbul ignore else
+            if (isMountedRef.current) {
+              setFormattedValue(lastFormattedValueRef.current);
+              setHasBadInput(false);
+            }
+          }
+        }
+      } else {
+        setHasBadInput(true);
+      }
+    }, [formatValue, onChange, parseString]);
+
+    const handleOnBlur = React.useCallback((event: React.FocusEvent<HTMLInputElement>) => {
+      updateQuantityValueFromString(event.target.value);
+    }, [updateQuantityValueFromString]);
+
+    function onInputKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {

```

Comment:
Unify event handler names as currently they are all different: `handleInputChange` (_Input_ w/o _On_), `handleOnBlur` (_On_ w/o _Input_) and `onInputKeyDown` (_Input_ w/o _On_, but with _on_ prefix).
```suggestion
    function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
```